### ISSUE TITLE
Potential python Pool memory leak

### ISSUE THREAD

**Issue by merrysailor** [catboost/catboost#2517]
catboost version: 1.2.2
Operating System: Rocky 8.5
CPU: Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
GPU: N/A

Hi,

I would like to bring up the following two examples. First:

```
import os
import numpy as np
import psutil
import catboost

X = np.empty((20_000_000, 128), order='F', dtype=np.float32)

for idx in range(X.shape[1]):
    X[:, idx] = np.random.random_sample(size=X.shape[0])

print('1', psutil.Process().memory_full_info().rss // 1_048_576)
pool = catboost.Pool(data=X)
print('2', psutil.Process().memory_full_info().rss // 1_048_576)
pool.quantize()
print('3', psutil.Process().memory_full_info().rss // 1_048_576)
del pool
print('4', psutil.Process().memory_full_info().rss // 1_048_576)
```

The output is:

```
1 9949
2 9950
3 13376
4 13376
```

The pool after quantization should be about 2.5GB. Is it reasonable to see an extra ~1.5GB in memory consumption?

Next consider if in addition we save the quantized pool:

```
import os
import numpy as np
import psutil
import catboost

X = np.empty((20_000_000, 128), order='F', dtype=np.float32)

for idx in range(X.shape[1]):
    X[:, idx] = np.random.random_sample(size=X.shape[0])

print('1', psutil.Process().memory_full_info().rss // 1_048_576)
pool = catboost.Pool(data=X)
print('2', psutil.Process().memory_full_info().rss // 1_048_576)
pool.quantize()
print('3', psutil.Process().memory_full_info().rss // 1_048_576)
pool.save('/tmp/pool.bin')
print('4', psutil.Process().memory_full_info().rss // 1_048_576)
```

The output is:
```
1 9950
2 9951
3 13341
4 17722
```

Is it reasonable to see another 4.5GB allocated and persisting in order to save the pool?

Thank you, please let me if I can provide more info.

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
@merrysailor yes, this should be reasonable since catboost caches allocated memory
you may try to create/destroy pool few thousand times in a loop to check that this behavior is because of caching

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
https://github.com/search?q=repo%3Acatboost%2Fcatboost+cpp-malloc-mimalloc&type=code

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
@andrey-khropov could you please help @merrysailor to create a custom build of catboost with system malloc/free instead of mimalloc?

**Comment by merrysailor** [comment]
Hello @Evgueni-Petrov-aka-espetrov ,

Thank you for the suggestion, I was able to build against the system_allocator and tcmalloc. I verified via 'strings' & MIMALLOC_SHOW_STATS=1.

It looks to me that mimalloc use is only part of the story. I added a couple more lines (3a, 4a) to further print max rss usage (resource.getrusage(resource.RUSAGE_SELF).ru_maxrss // 1_024).

This is how things look with tcmalloc:
1 9882
2 9883
3 12636
3a 12714
4 15721
4a 17520

And with the system_allocator:
1 9882
2 9884
3 12579
3a 12715
4 15659
4a 17459

It looks that even with the system allocator, there is 
a) Memory usage to save the Pool that is proportional to its size, which seesm odd
b) Memory that is not freed.

Thank you!
